### PR TITLE
Add more sysvar API docs

### PR DESCRIPTION
**Pull request by brson**
#### Problem

More docs needed.

#### Summary of Changes

This adds documentation to the sysvar module and its submodules, to the data types and supporting functions associated with each sysvar. It moves the crate-level sysvar docs into the sysvar module.

### CODE REVIEW

**Comment by brson** [comment]
The CI sanity check is tripping on `println!` invocations in examples that are written correctly. I am willing to change them to `msg!` but it will not be idiomatic.

**Comment by brson** [comment]
I'll think about other ways to present those examples.

**Comment by CriesofCarrots** [comment]
> The CI sanity check is tripping on `println!` invocations in examples that are written correctly. I am willing to change them to `msg!` but it will not be idiomatic.

Ah, yeah, our println net is too broad. We ought to allow them in docs. I'll see how tedious that would be to update.

**Comment by brson** [comment]
@CriesofCarrots i just removed the printlns from the examples

File: sdk/program/src/clock.rs [hunk 1]
```diff
@@ -1,20 +1,45 @@
 //! Information about the network's clock, ticks, slots, etc.
+//!
+//! Time in Solana is marked primily by _slots_, which occur approximately every
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:3]:
```suggestion
//! Time in Solana is marked primarily by _slots_, which occur approximately every
```

File: sdk/program/src/clock.rs [hunk 2]
```diff
@@ -1,20 +1,45 @@
...
+//! 400 milliseconds, and are numbered sequentially. For every slot a leader is
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:4]:
```suggestion
//! 400 milliseconds, and are numbered sequentially. For every slot, a leader is
```

File: sdk/program/src/clock.rs [hunk 3]
```diff
@@ -1,20 +1,45 @@
...
+//! chosen from the validator set, and that leader is expected to produce a new
+//! block, though sometimes leaders may fail to do so. Blocks can be identified
+//! by their slot number, and some slots do not contain a block.
+//!
+//! An approximation of the passage of real-world time can be calculated by
+//! multiplying a number of slots by [`SLOT_MS`], which is a constant target
+//! time for the network to produce slots. Note though that this method suffers
+//! a variable amount of drift, as the network does not produce slots at exactly
+//! the target rate, and the greater number of slots being calculated for, the
+//! greater the drift. Epochs can not be used this way as they contain variable
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:14]:
```suggestion
//! greater the drift. Epochs cannot be used this way as they contain variable
```

File: sdk/program/src/clock.rs [hunk 4]
```diff
@@ -1,20 +1,45 @@
...
 #[cfg(test)]
 static_assertions::const_assert_eq!(MS_PER_TICK, 6);
+
+/// The number of milliseconds per tick (6).
 pub const MS_PER_TICK: u64 = 1000 / DEFAULT_TICKS_PER_SECOND;
 
 #[cfg(test)]
 static_assertions::const_assert_eq!(SLOT_MS, 400);
+
+/// The duration of a slot (400 milliseconds).
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:42]:
```suggestion
/// The expected duration of a slot (400 milliseconds).
```
Thoughts?

File: sdk/program/src/clock.rs [hunk 5]
```diff
@@ -52,12 +80,14 @@ static_assertions::const_assert_eq!(DEFAULT_MS_PER_SLOT, 400);
...
 pub const DEFAULT_S_PER_SLOT: f64 = DEFAULT_TICKS_PER_SLOT as f64 / DEFAULT_TICKS_PER_SECOND as f64;
 
-/// The time window of recent block hash values that the bank will track the signatures
-/// of over. Once the bank discards a block hash, it will reject any transactions that use
-/// that `recent_blockhash` in a transaction. Lowering this value reduces memory consumption,
-/// but requires clients to update its `recent_blockhash` more frequently. Raising the value
-/// lengthens the time a client must wait to be certain a missing transaction will
-/// not be processed by the network.
+/// The time window of recent block hash values that the bank will track the
+/// signatures of over.
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:84]:
```suggestion
/// The time window of recent block hash values over which the bank will track
/// signatures.
```

File: sdk/program/src/clock.rs [hunk 6]
```diff
@@ -52,12 +80,14 @@ static_assertions::const_assert_eq!(DEFAULT_MS_PER_SLOT, 400);
...
+///
+/// Once the bank discards a block hash, it will reject any transactions that
+/// use that `recent_blockhash` in a transaction. Lowering this value reduces
+/// memory consumption, but requires clients to update its `recent_blockhash`
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:88]:
```suggestion
/// memory consumption, but requires a client to update its `recent_blockhash`
```

File: sdk/program/src/clock.rs [hunk 7]
```diff
@@ -78,52 +108,61 @@ pub const MAX_TRANSACTION_FORWARDING_DELAY_GPU: usize = 2;
...
-///  current Slot, as Epochs vary in duration (they start short and grow
-///  as the network progresses).
+/// A representation of network time.
 ///
+/// All members of `Clock` start from 0 upon network boot.
 #[repr(C)]
 #[derive(Serialize, Deserialize, Debug, Default, PartialEq, Eq)]
 pub struct Clock {
-    /// the current network/bank Slot
+    /// The current network/bank `Slot`.
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:148]: This predates your work, but it feels weird to me that these are defined in relation to a Bank, but this struct is itself a member of a Bank. Wdyt about removing references to bank here?

```suggestion
    /// The current `Slot`.
```

File: sdk/program/src/clock.rs [hunk 8]
```diff
@@ -78,52 +108,61 @@ pub const MAX_TRANSACTION_FORWARDING_DELAY_GPU: usize = 2;
...
     pub slot: Slot,
-    /// the timestamp of the first Slot in this Epoch
+    /// The timestamp of the first `Slot` in this `Epoch`.
     pub epoch_start_timestamp: UnixTimestamp,
-    /// the bank Epoch
+    /// The bank `Epoch`.
```
Comment (CriesofCarrots) [inline sdk/program/src/clock.rs:152]:
```suggestion
    /// The current `Epoch`.
```

File: sdk/program/src/epoch_schedule.rs [hunk 9]
```diff
@@ -1,23 +1,37 @@
 //! Configuration for epochs and slots.
+//!
+//! Epochs mark a period of time composed of _slots_, for which a particular
+//! [leader schedule][ls] is in effect. The epoch schedule determines the length
+//! of epochs, and the timing of the next leader schedule selection.
```
Comment (CriesofCarrots) [inline sdk/program/src/epoch_schedule.rs:5]:
```suggestion
//! of epochs, and the timing of the next leader-schedule selection.
```

File: sdk/program/src/epoch_schedule.rs [hunk 10]
```diff
@@ -1,23 +1,37 @@
...
+/// The default is an entire epoch, i.e. leader schedule for epoch X is calculated at
+/// the beginning of epoch X - 1.
 pub const DEFAULT_LEADER_SCHEDULE_SLOT_OFFSET: u64 = DEFAULT_SLOTS_PER_EPOCH;
 
 /// The maximum number of slots before an epoch starts to calculate the leader schedule.
-///  Default is an entire epoch, i.e. leader schedule for epoch X is calculated at
-///  the beginning of epoch X - 1.
+///
+/// Default is an entire epoch, i.e. leader schedule for epoch X is calculated at
+/// the beginning of epoch X - 1.
```
Comment (CriesofCarrots) [inline sdk/program/src/epoch_schedule.rs:29]: Maybe just remove this?

File: sdk/program/src/sysvar/mod.rs [hunk 11]
```diff
@@ -1,5 +1,83 @@
...
+//! overhead of deserialization while also requiring the sysvar account address
+//! be passed to the program, wasting the limited space available to
+//! transactions. Deserializing sysvars that can instead be retrieved with
+//! `Sysvar::get` should be only be considered for compatibility with older
+//! programs that pass around sysvar accounts.
+//!
+//! Some sysvars are too large to deserialize within a program, and
+//! `Sysvar::from_account_info` returns an error. Some sysvars are too large
+//! to deserialize within a program, and attempting to will exhaust the
+//! program's compute budget. Some sysvars do not implement `Sysvar::get` and
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/mod.rs:72]:
```suggestion
//! Some sysvars are too large to deserialize within a program, and
//! `Sysvar::from_account_info` returns an error, or the serialization attempt 
//! will exhaust the program's compute budget. 
//! Some sysvars do not implement `Sysvar::get` and
```
To be a little less repetitive

File: sdk/program/src/sysvar/recent_blockhashes.rs [hunk 12]
```diff
@@ -1,3 +1,21 @@
+//! Information about the network’s clock, ticks, slots, etc.
+//!
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/recent_blockhashes.rs:2]: Oops, this is from clock. Update or remove?

File: sdk/program/src/rent.rs [hunk 13]
```diff
@@ -3,40 +3,54 @@
...
+///
+/// Valid values are in the range [0, 100]. The remaining percentage is
+/// distributed to validators.
 pub const DEFAULT_BURN_PERCENT: u8 = 50;
 
-/// account storage overhead for calculation of base rent
+/// Account storage overhead for calculation of base rent.
+///
+/// This is an approximate constant number of bytes required to store an account
+/// of any size. It is added to an accounts data length when calculating
```
Comment (CriesofCarrots) [inline sdk/program/src/rent.rs:52]: Actually I believe this is the exact number of bytes required for an account with no data.

File: sdk/program/src/sysvar/clock.rs [hunk 14]
```diff
@@ -1,5 +1,131 @@
...
+//! # let accounts = &[a.clone(), a];
+//! # process_instruction(
+//! #     &Pubkey::new_unique(),
+//! #     accounts,
+//! #     &[],
+//! # )?;
+//! # Ok::<(), ProgramError>(())
+//! ```
+//!
+//! Accessing via the RPC client:
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/clock.rs:98]: Ah, nice idea

File: sdk/program/src/sysvar/slot_hashes.rs [hunk 15]
```diff
@@ -1,7 +1,50 @@
-//! named accounts for synthesized data accounts for bank state, etc.
+//! The most recent hashes of a slot's parent banks.
 //!
-//! this account carries the Bank's most recent bank hashes for some N parents
+//! The _slot hashes sysvar_ provides access to the [`SlotHashes`] type.
 //!
+//! The [`Sysvar::from_account_info`] and [`Sysvar::get`] methods always returns
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/slot_hashes.rs:5]:
```suggestion
//! The [`Sysvar::from_account_info`] and [`Sysvar::get`] methods always return
```

File: sdk/program/src/sysvar/slot_history.rs [hunk 16]
```diff
@@ -1,8 +1,52 @@
-//! named accounts for synthesized data accounts for bank state, etc.
+//! Access to slots present over the last epoch.
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/slot_history.rs:1]:
```suggestion
//! A bitvector of slots present over the last epoch.
```
I'm not set on my suggested wording, but "access" applies something bigger to me

File: sdk/program/src/sysvar/slot_history.rs [hunk 17]
```diff
@@ -1,8 +1,52 @@
...
 //!
-//! this account carries a bitvector of slots present over the past
-//! epoch
+//! The _slot history sysvar_ provides access to the [`SlotHistory`] type.
 //!
+//! The [`Sysvar::from_account_info`] and [`Sysvar::get`] methods always returns
```
Comment (CriesofCarrots) [inline sdk/program/src/sysvar/slot_history.rs:5]:
```suggestion
//! The [`Sysvar::from_account_info`] and [`Sysvar::get`] methods always return
```

**Review by CriesofCarrots — COMMENTED** [review]
Some nits, but generally great work, as always!

File: sdk/program/src/rent.rs (hunk 13 above)
Comment (brson) [reply sdk/program/src/rent.rs]: I've changed it. Can't remember what I was thinking with "approximate".

File: sdk/program/src/epoch_schedule.rs (hunk 10 above)
Comment (brson) [reply sdk/program/src/epoch_schedule.rs:26]: Done.

**Comment by brson** [comment]
Great reviews, thanks.

**Comment by brson** [comment]
CI failure is a rustsec advisory unrelated to this patch.

**Review by CriesofCarrots — DISMISSED** [review]
Lgtm, thank you!
I'm going to rebase on  master to pick up the audit fix and make sure we get through `checks`

**Pull request merged**